client/src/js/services/FilterService.js

```diff
@@ -4,50 +4,183 @@ angular.module('bhima.services')
 FilterService.$inject = ['Store'];
 
 function FilterService(Store) {
-  // @FIXME this should be defined as a constant accross the server and the client
-  /* @const */
-  var PERIODS = new Store({ identifier: 'key' });
-  PERIODS.setData([
-    {
-      key   : 'today',
-      label : 'FORM.LABELS.TODAY',
-    },{
-      key   : 'week',
-      label : 'FORM.LABELS.THIS_WEEK',
-    },{
-      key   : 'month',
-      label : 'FORM.LABELS.THIS_MONTH',
-    },
-  ]);
-  var DEFAULT_PERIOD = 'today';
-
-  function Filter() {
-    this.defaultPeriod = PERIODS.get(DEFAULT_PERIOD);
-  }
-
-  Filter.prototype.applyDefaults = function applyDefaults(filters) {
-    if (!filters) { return filters; }
-
-    var emptyFilters = Object.keys(filters).length === 0;
-
-    if (emptyFilters) {
-      filters.defaultPeriod = this.defaultPeriod.key;
+
+  function FilterList() {
+    // initialise internal state
+    this._defaultFilters = [];
+    this._customFilters = [];
+
+    this._filterIndex = {};
+  }
+
+  FilterList.prototype.resetFilterState = function resetFilterState(key) {
+    this._filterIndex[key].assign(null, null);
+  }
+
+  FilterList.prototype._resetCustomFilters = function resetCustomFilters() {
+    this._filterActiveFilters().forEach(function (filter) {
+      // only by default remove custom values
+      if (!filter._isDefault) {
+        this.resetFilterState(filter._key);
+      }
+    }.bind(this));
+  }
+
+  // @TODO registerDefaultFilter and registerCustomFilter could use the same underlying function
+  //       with a toggle between the array to populate and the default value
+  FilterList.prototype.registerDefaultFilters = function registerDefaultFilters(filterDefinitions) {
+    var formattedFilters = filterDefinitions.map(function (filterDefinition) {
+      var filter = new Filter(filterDefinition.key, filterDefinition.label, filterDefinition.valueFilter);
+      filter.setDefault(true);
+
+      if (filterDefinition.defaultValue) {
+        filter.assign(filterDefinition.defaultValue);
+      }
+      return filter;
+    });
+
+    // udpate index
+    this._indexList(this._filterIndex, formattedFilters);
+    this._defaultFilters = this._defaultFilters.concat(formattedFilters);
+  };
+
+  FilterList.prototype.registerCustomFilters = function registerCustomFilters(filterDefinitions) {
+    var formattedFilters = filterDefinitions.map(function (filterDefinition) {
+      var filter = new Filter(filterDefinition.key, filterDefinition.label, filterDefinition.valueFilter);
+      filter.setDefault(false);
```

Comment:
I would have no problem with this refactor - I think it can stay explicit for now and on the next round of development the internal methods can be streamlined.